File: leetcode/java/0435. Non-overlapping Intervals.java
package leetcode.java;

// Time complexity: O(NlogN)
// Space complexity: O(1)

import java.util.Arrays;

class Solution {
    public int eraseOverlapIntervals(int[][] intervals) {
        Arrays.sort(intervals, (a, b) -> a[0] - b[0]);
        int toRemove = 0;
        int prev = -999999;

        for (int[] interval : intervals) {
            if (prev == -999999) {
                prev = interval[1];
            } else if (prev > interval[0]) {
                toRemove += 1;
                prev = Integer.min(interval[1], prev);
            } else {
                prev = interval[1];
            }
        }
        return toRemove;
    }
}

File: leetcode/python/0435. Non-overlapping Intervals.py
# Time complexity: O(NlogN)
# Space complexity: O(1)

class Solution(object):
    def eraseOverlapIntervals(self, intervals):
        
        intervals.sort(lambda x: x[0])
        changes = 0
        prev = intervals[0][1]

        for start, end in intervals[1:]:
            if start >= prev:
                prev = end
            else:
                changes += 1
                prev = min(prev, end)
        return changes